process/sync/metablock.go

```diff
@@ -125,74 +125,178 @@ func (boot *MetaBootstrap) syncFromStorer(
-) error {
-	return nil
+	for i := uint32(0); i < boot.shardCoordinator.NumberOfShards(); i++ {
+		if nonceWithLastNotarized[i]-nonceWithFinalNotarized[i] > 1 {
+			finalNotarized[i] = lastNotarized[i]
+		}
+
+		log.Info(fmt.Sprintf("last notarized block from shard %d is %d and final notarized block is %d\n",
+			i, lastNotarized[i], finalNotarized[i]))
+	}
+
+	return startNonce, finalNotarized, lastNotarized
 }
 
-func (boot *MetaBootstrap) applyNotarizedBlock(
-	nonce uint64,
-	notarizedHdrNonceHashDataUnit dataRetriever.UnitType,
+func (boot *MetaBootstrap) applyNotarizedBlocks(
+	finalNotarized map[uint32]uint64,
+	lastNotarized map[uint32]uint64,
 ) error {
-	nonceToByteSlice := boot.uint64Converter.ToByteSlice(nonce)
-	headerHash, err := boot.store.Get(notarizedHdrNonceHashDataUnit, nonceToByteSlice)
-	if err != nil {
-		return err
-	}
+	for i := uint32(0); i < boot.shardCoordinator.NumberOfShards(); i++ {
+		nonce := finalNotarized[i]
+		headerHandler, _, err := boot.getShardHeaderFromStorage(i, nonce)
+		if err != nil {
+			return err
+		}
 
-	header, err := process.GetShardHeaderFromStorage(headerHash, boot.marshalizer, boot.store)
-	if err != nil {
-		return err
-	}
+		boot.blkExecutor.SetLastNotarizedHdr(i, headerHandler)
 
-	log.Info(fmt.Sprintf("apply notarized block with nonce %d and round %d\n", header.Nonce, header.Round))
+		nonce = lastNotarized[i]
+		headerHandler, _, err = boot.getShardHeaderFromStorage(i, nonce)
+		if err != nil {
+			return err
+		}
 
-	if header.GetRound() > boot.boostrapRoundIndex {
-		return ErrHigherBootstrapRound
+		boot.blkExecutor.SetLastNotarizedHdr(i, headerHandler)
 	}
 
-	boot.blkExecutor.SetLastNotarizedHdr(header.ShardId, header)
-
 	return nil
 }
 
-func (boot *MetaBootstrap) getHeaderFromStorage(nonce uint64) (data.HeaderHandler, []byte, error) {
-	nonceToByteSlice := boot.uint64Converter.ToByteSlice(nonce)
-	headerHash, err := boot.store.Get(dataRetriever.MetaHdrNonceHashDataUnit, nonceToByteSlice)
-	if err != nil {
-		return nil, nil, err
-	}
+func (boot *MetaBootstrap) cleanupNotarizedStorage(lastNotarized map[uint32]uint64) {
+	for i := uint32(0); i < boot.shardCoordinator.NumberOfShards(); i++ {
+		hdrNonceHashDataUnit := dataRetriever.ShardHdrNonceHashDataUnit + dataRetriever.UnitType(i)
+		highestNonceInStorer := boot.computeHighestNonce(hdrNonceHashDataUnit)
 
-	header, err := process.GetMetaHeaderFromStorage(headerHash, boot.marshalizer, boot.store)
+		for i := lastNotarized[i] + 1; i <= highestNonceInStorer; i++ {
+			errNotCritical := boot.removeBlockHeader(i, dataRetriever.BlockHeaderUnit, hdrNonceHashDataUnit)
+			if errNotCritical != nil {
+				log.Info(fmt.Sprintf("remove notarized block header with nonce %d: %s\n", i, errNotCritical.Error()))
+			}
+		}
+	}
+}
 
-	return header, headerHash, err
+func (boot *MetaBootstrap) removeBlockBody(
+	nonce uint64,
+	blockUnit dataRetriever.UnitType,
+	hdrNonceHashDataUnit dataRetriever.UnitType,
+) error {
+	return nil
```

Comment:
function does not do anything. delete

Reply:
It should implement the interface now, so it should remain like this (for meta we do not need to delete block body but the method should exist to implement the interface)